### PR TITLE
Support float config on server side

### PR DESCRIPTION
**Pull request by 0xfaulty**
Redo of #5894 PR with nan/inf check and map settings support. 

Perhaps this type of configuration won't be used in DDNet, but for modifications, it's an extremely useful feature. Before adding support for float configurations, I had to specify values as percentages or divide them by some constant. Yes, in some cases the precision might not be perfect, but for most of my tasks, this is more than sufficient.

To prevent breaking values by re-saving, I suggest using these configurations **only for the server side** where we only read without write.

**What's the difficulty in saving as an int and divide/multiplying by 1000?**
The problem lies in the fact that the configuration file is loaded via CConsole::ExecuteFile, meaning all lines are treated as console inputs. As a result, 
1. either a custom implementation for reading configurations needs to be written, 
2. or the console must be made aware that it's reading from a file, 
3. or int values will also need to be entered in the console. 

In any case, this leads to either bad code or a massive refactor, which could potentially be done later [^1] if someone needs to add a configuration variable specifically on the client side.  For the server side, I find the current implementation fully sufficient.

[^1]: The current version won't need to be changed in any case, as it specifically parses console input, and the input into the console will be float values

## Checklist

- [ ] Tested the change ingame
- [ ] Provided screenshots if it is a visual change
- [ ] Tested in combination with possibly related configuration options
- [ ] Written a unit test (especially base/) or added coverage to integration test
- [ ] Considered possible null pointers and out of bounds array indexing
- [ ] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

### CODE REVIEW

**Comment by heinrich5991** [comment]
I suggest this is done instead using decimals. I.e. allow the user to input numbers like `12.345`, have a configured number of decimals, e.g. 4, and then save the number as 123450. This allows us to side-step all problems of floating-point numbers, including different compiler behavior, roundtripping, etc.

Is there a downside to this approach?

**Comment by 0xfaulty** [comment]
> I suggest this is done instead using decimals. I.e. allow the user to input numbers like `12.345`, have a configured number of significant digits, e.g. 4, and then save the number as 123450. This allows us to side-step all problems of floating-point numbers, including different compiler behavior, roundtripping, etc.
> 
> Is there a downside to this approach?

@heinrich5991 I only need to change one line of code to serialize it as an int, but rewriting or duplicating the entire configuration reading logic to handle decimals seems excessive. Could you please review the code and let me know if there’s a way to achieve this without duplicating the logic in console.cpp? (It’s possible I’m overlooking something.)

[client.cpp:4774](https://github.com/ddnet/ddnet/blob/010d7ce1fc123929f765229536be50efce5a5c75/src/engine/client/client.cpp#L4774) pConsole->ExecuteFile(CONFIG_FILE) // read config
[console.cpp:660](https://github.com/ddnet/ddnet/blob/010d7ce1fc123929f765229536be50efce5a5c75/src/engine/shared/console.cpp#L660) ExecuteLine(pLine, ClientId) // each line excecuted as command

Also as I mentioned in my initial message, I’m not sure this code is needed on the client side at all. On the server side, server owners fill out configurations without using the client, and I don’t think they would want to write tunings as regular floats while handling other configurations in a different way.

In other words, I assume the current implementation will remain for the server, and any other implementation can be added to the client side whenever someone actually needs it.

**Comment by heinrich5991** [comment]
> On the server side, server owners fill out configurations without using the client, and I don’t think they would want to write tunings as regular floats while handling other configurations in a different way.

Tunings are also internally saved as decimals (with one decimal digit). I'm suggesting that you accept inputs like "12.345" and only internally store them as fixed-point numbers.

> I only need to change one line of code to serialize it as an int, but rewriting or duplicating the entire configuration reading logic to handle decimals seems excessive. Could you please review the code and let me know if there’s a way to achieve this without duplicating the logic in console.cpp? (It’s possible I’m overlooking something.)

I don't understand this. Parsing a decimal is parsing everything in front of the comma and everything afterwards. I don't understand how the code fragments you link to are related to that. The code fragments you linked are about high-level execution, but I'd like you to change parameter handling instead.

**Comment by Jupeyy** [comment]
> Is there a downside to this approach?

You break all config if you ever increase the accuracy of whatever float/fixed-point you use if you save them as integers.
Even tho I generally agree floats suck and compiler difference suck as well, a 64-bit double can probs represent anything a _normal_ human inputs and as soon as you input something that is beyond that limit the 32-bit integer representation breaks more than the 64-bit float representation ^^

@heinrich5991 do you just generally dislike to add another variant to config files, or do you actually fear about rounding errors? Like can you give an example of which number would break while it doesn't using 32-bit fixed-points?

**Comment by 0xfaulty** [comment]
> I don't understand this. Parsing a decimal is parsing everything in front of the comma and everything afterwards. I don't understand how the code fragments you link to are related to that. The code fragments you linked are about high-level execution, but I'd like you to change parameter handling instead.

@heinrich5991 This is exactly what I’m trying to explain: the same high-level approach is currently used for both user input processing and file reading through ExecuteLine, which simply takes a string as argument. As a result, if we "internally store" the float `100.123` as an int `100123`, it will be interpreted as the float `100123`[^1] when reading the configuration, which will then be saved as `100123000`, and so on. Due to this shared high-level approach, I don’t see a good way to change the reading behavior specifically for file inputs.

Here are the call stack traces to make it easier for you to compare:

image

[^1]: Because at this point, we don’t know whether it was file reading or user input processing.

**Comment by heinrich5991** [comment]
> You break all config if you ever increase the accuracy of whatever float/fixed-point you use if you save them as integers.
> Even tho I generally agree floats suck and compiler difference suck as well, a 64-bit double can probs represent anything a _normal_ human inputs and as soon as you input something that is beyond that limit the 32-bit integer representation breaks more than the 64-bit float representation ^^

I think I was unclear.

I want to save the fixed points to something like `cl_zoom 1.25` to the config file. Just when parsing and saving it inside the running client itself, I'd like it to be stored into a fixed-point integer with a given number of decimal digits.

> @heinrich5991 do you just generally dislike to add another variant to config files, or do you actually fear about rounding/precision errors? Like can you give an example of which number would break while it doesn't using 32-bit fixed-points?

I fear that you put something like 1.4 into the config and it gets saved as 1.39999999.

**Comment by Kaffeine** [comment]
Just in a case, here is implementation for Infclass server: https://github.com/infclass/teeworlds-infclassR/commit/f57c626847c367637d22045a4d13ae2ec9f8f653
It uses fixed-point number as storage.

**Comment by Kaffeine** [comment]
Ticket for this feature: https://github.com/ddnet/ddnet/issues/7321

**Comment by ChillerDragon** [comment]
This is now draft since 9 months without progress. @0xfaulty are you still interested in this? Are you planning to work on it?

**Comment by 0xfaulty** [comment]
> This is now draft since 9 months without progress. @0xfaulty are you still interested in this? Are you planning to work on it?

I'm still interesting in adding it to ddnet, but I'm afraid of facing criticism of the new version. I think I could try to prepare version from infclass tommorow, and let's see..

File: src/engine/shared/fixed_point_number.cpp [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+	const int Whole = m_Value / 1000;
+	int Fractional = std::abs(m_Value) % 1000;
+	auto result = std::to_chars(s_aBuf, &s_aBuf[c_FixedPointStrBufferSize - 1], Whole);
+	if(Fractional)
+	{
+		result.ptr[0] = '.';
+		while(Fractional % 10 == 0)
+			Fractional /= 10;
+
+		result = std::to_chars(&result.ptr[1], &s_aBuf[c_FixedPointStrBufferSize - 1], Fractional);
```
Comment (Kaffeine) [inline src/engine/shared/fixed_point_number.cpp]: I suspect a bug here. The code chops the trailing zeroes but probably it misses (does not write to the result) the leading zeros.
As I read it now, the code 'll print `.1` for any of `{Fractional = 100, Fractional = 10, Fractional = 1}`.

**Comment by Kaffeine** [comment]
I think it would be fair for me to ask for a [Co-authored-by](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors) trailer.

**Comment by ChillerDragon** [comment]
@Kaffeine could you send the exact co author string you want so @0xfaulty does not have to guess

**Comment by Kaffeine** [comment]
```
$ git log f57c626847c367637d22045a4d13ae2ec9f8f653
commit f57c626847c367637d22045a4d13ae2ec9f8f653
Author: Alexander Akulich <akulichalexander@gmail.com>
```
so
```
Co-authored-by: Alexander Akulich <akulichalexander@gmail.com>
```

**Comment by 0xfaulty** [comment]
Sure, but it's still not finished yet, I will check later on pc, on laptop impossible everything lagg to much :(

**Comment by 0xfaulty** [comment]
@heinrich5991 could you please re-check PR?

File: src/engine/shared/fixed_point_number.cpp (hunk 1 above)
Comment (0xfaulty) [reply src/engine/shared/fixed_point_number.cpp]: rewrote this part

**Comment by heinrich5991** [comment]
I still see a lot of `float`s in that diff. The idea of using fixed point numbers is that you don't round-trip through `float`s, as that's where the roundings happen.

You can parse the fixed-point numbers yourself by looking for the decimal dot and using integer parsing.

**Comment by 0xfaulty** [comment]
@heinrich5991 reworked, please check again

File: src/base/fixed.h [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+#define ENGINE_SHARED_FIXED_POINT_NUMBER_H
+
+#include <cstddef>
+#include <cstdint>
+#include <limits>
+
+class CFixedPointNumber
+{
+public:
+	using Underlying = int32_t;
```
Comment (heinrich5991) [inline src/base/fixed.h:10]: Make this private.

File: src/base/fixed.h [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+	// Fixed-point format: 3 decimal places -> SCALE = 1000.
+	static constexpr Underlying SCALE = 1000;
+	static constexpr int NUM_DECIMAL_DIGITS = 3;
```
Comment (heinrich5991) [inline src/base/fixed.h:14]: Private.

File: src/engine/shared/fixed_point_number.h [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+	static bool FromStr(const char *pStr, CFixedPointNumber &Out);
+
+	// Compile-time parse for string literals.
+	template<size_t N>
+	static consteval CFixedPointNumber FromLiteral(const char (&Lit)[N])
+	{
+		Underlying V{};
+		if(!ParseLiteral(Lit, N, V))
+		{
+			return CFixedPointNumber(0);
```
Comment (heinrich5991) [inline src/engine/shared/fixed_point_number.h]: Can we throw here, instead?

File: src/engine/shared/fixed_point_number.h [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+		}
+		return FromRaw(V);
+	}
+
+	// Convert to C-string. Returns a pointer to a thread_local buffer.
```
Comment (heinrich5991) [inline src/engine/shared/fixed_point_number.h]: This is not a pattern we use in DDNet. It's also error-prone, so please pass in a buffer instead.

File: src/engine/shared/fixed_point_number.h [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+	const char *AsStr() const;
+
+	// Raw access and conversions.
+	static constexpr CFixedPointNumber FromRaw(const Underlying V) { return CFixedPointNumber(V); }
```
Comment (heinrich5991) [inline src/engine/shared/fixed_point_number.h]: Private.

File: src/base/fixed.h [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+
+	// Convert to C-string. Returns a pointer to a thread_local buffer.
+	const char *AsStr() const;
+
+	// Raw access and conversions.
+	static constexpr CFixedPointNumber FromRaw(const Underlying V) { return CFixedPointNumber(V); }
+	constexpr bool IsZero() const { return m_Value == 0; }
+	constexpr explicit operator float() const { return static_cast<float>(m_Value) / static_cast<float>(SCALE); }
+
+	// Comparisons.
```
Comment (heinrich5991) [inline src/base/fixed.h:42]: Are these comments auto-generated?

File: src/engine/shared/fixed_point_number.h [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+
+	// Comparisons.
+	constexpr bool operator==(const CFixedPointNumber &O) const { return m_Value == O.m_Value; }
+	constexpr bool operator!=(const CFixedPointNumber &O) const { return m_Value != O.m_Value; }
+	constexpr bool operator<(const CFixedPointNumber &O) const { return m_Value < O.m_Value; }
+	constexpr bool operator<=(const CFixedPointNumber &O) const { return m_Value <= O.m_Value; }
+	constexpr bool operator>(const CFixedPointNumber &O) const { return m_Value > O.m_Value; }
+	constexpr bool operator>=(const CFixedPointNumber &O) const { return m_Value >= O.m_Value; }
+
+	constexpr CFixedPointNumber() = default;
```
Comment (heinrich5991) [inline src/engine/shared/fixed_point_number.h]: Perhaps another `CFixedPointNumber(int)` constructor that takes an integer and converts it into the corresponding fixed point number.

File: src/engine/shared/fixed_point_number.h [hunk 9]
```diff
@@ -0,0 +1,173 @@
+#ifndef ENGINE_SHARED_FIXED_POINT_NUMBER_H
+#define ENGINE_SHARED_FIXED_POINT_NUMBER_H
+
+#include <cstddef>
+#include <cstdint>
+#include <limits>
+
+class CFixedPointNumber
```
Comment (heinrich5991) [inline src/engine/shared/fixed_point_number.h]: Name is quite long. Perhaps it could be named `fixed` and moved to src/base?

File: src/game/editor/editor_server_settings.cpp [hunk 10]
```diff
@@ -44,6 +45,15 @@ struct SMapSettingInt : public IMapSetting
 	SMapSettingInt(const char *pName, const char *pHelp, int Default, int Min, int Max) :
 		IMapSetting(pName, pHelp, IMapSetting::SETTING_INT), m_Default(Default), m_Min(Min), m_Max(Max) {}
 };
+struct SMapSettingFloat : public IMapSetting
```
Comment (heinrich5991) [inline src/game/editor/editor_server_settings.cpp]:
```suggestion
struct SMapSettingFixed : public IMapSetting
```

File: src/game/server/teehistorian.cpp [hunk 11]
```diff
@@ -172,6 +172,20 @@ void CTeeHistorian::WriteHeader(const CGameInfo *pGameInfo)
 		First = false; \
 	}
 
+#define MACRO_CONFIG_FLOAT(Name, ScriptName, Def, Min, Max, Flags, Desc) \
```
Comment (heinrich5991) [inline src/game/server/teehistorian.cpp]:
```suggestion
#define MACRO_CONFIG_FIXED(Name, ScriptName, Def, Min, Max, Flags, Desc) \
```

File: src/base/fixed.h (hunk 2 above)
Comment (0xfaulty) [reply src/base/fixed.h:10]: done

File: src/base/fixed.h (hunk 3 above)
Comment (0xfaulty) [reply src/base/fixed.h:14]: done

File: src/engine/shared/fixed_point_number.h (hunk 5 above)
Comment (0xfaulty) [reply src/engine/shared/fixed_point_number.h]: rewrote to a buffer approach

File: src/engine/shared/fixed_point_number.h (hunk 6 above)
Comment (0xfaulty) [reply src/engine/shared/fixed_point_number.h]: done

File: src/base/fixed.h (hunk 7 above)
Comment (0xfaulty) [reply src/base/fixed.h:42]: yes, I asked generate comments that could simplify review, I can delete them if they are redundant

File: src/engine/shared/fixed_point_number.h (hunk 8 above)
Comment (0xfaulty) [reply src/engine/shared/fixed_point_number.h]: added FromInt static function otherwise will conflict with private constructor

File: src/engine/shared/fixed_point_number.h (hunk 9 above)
Comment (0xfaulty) [reply src/engine/shared/fixed_point_number.h]: renamed to `CFixed`, as according to the code style description in the project, classes should be named with the prefix C. If you mean to literally rename it to `fixed` as a primitive type, let me know

File: src/game/editor/editor_server_settings.cpp (hunk 10 above)
Comment (0xfaulty) [reply src/game/editor/editor_server_settings.cpp]: renamed

File: src/game/server/teehistorian.cpp (hunk 11 above)
Comment (0xfaulty) [reply src/game/server/teehistorian.cpp]: renamed

**Comment by 0xfaulty** [comment]
I had to rebase the branch because clang-format changed in the master branch

File: src/engine/shared/fixed_point_number.h (hunk 4 above)
Comment (0xfaulty) [reply src/engine/shared/fixed_point_number.h]: I don’t see a clean solution. Options so far:
1. throw: not allowed in this project
2. static_assert: only works if added to every call site, which is messy/ambiguous
3. Crash tactics (div/0, OOB, null): I’d rather avoid these

**Comment by 0xfaulty** [comment]
@heinrich5991 Please check again, I resolved everything except [throw comment](https://github.com/ddnet/ddnet/pull/9288#discussion_r2513507698). If the current approach doesn’t work for you and you have time, I’d be very grateful if you could suggest an alternative solution.

File: src/engine/shared/fixed_point_number.h (hunk 9 above)
Comment (Robyt3) [reply src/engine/shared/fixed_point_number.h]: `CFixed` is a bit too vague to me. I don't know how to make it much shorter than `CFixedPointNumber` though.

I suppose our existing functions like `f2fx` aren't much clearer either though.

File: src/engine/shared/config.cpp [hunk 12]
```diff
@@ -109,6 +109,90 @@ void SIntConfigVariable::ResetToOld()
...
+		pData->m_pVariable->AsStr(aValueStr, sizeof(aValueStr));
+		char aBuf[64];
+		str_format(aBuf, sizeof(aBuf), "Value: %s", aValueStr);
+		pData->m_pConsole->Print(IConsole::OUTPUT_LEVEL_STANDARD, "config", aBuf);
+	}
+}
+
+void SFixedConfigVariable::Register()
+{
+	m_pConsole->Register(m_pScriptName, "?s", m_Flags, CommandCallback, this, m_pHelp);
```
Comment (Robyt3) [inline src/engine/shared/config.cpp:155]: Why does this register the config variable as a command with string argument? This should be integrated into console argument parsing to also support commands with fixed point number arguments.

Comment (0xfaulty) [reply src/engine/shared/config.cpp:155]: Are you suggesting adding some new literal for this? I used a string because 'f' cannot be used due to conversion, and there are no other options at the moment.

Comment (Robyt3) [reply src/engine/shared/config.cpp:155]: Yes. Config variables are just console commands. The arguments of commands should be parsed in the console directly. This otherwise adds a lot of duplicate code. The editor code you are adding here already incorrectly tries to represent map settings with fixed point arguments as commands with `f` argument and assumes they are fixed point.

Why do we even have a separate `f` argument for `float`s? Can we repurpose that argument for fixed point numbers and remove float arguments or do the fixed point numbers as of this PR not cover all use cases of existing float arguments?
